### PR TITLE
Update Independent samples

### PR DESCRIPTION
**Pull request by komalsrathi**
#### What scientific question is your analysis addressing?

Selection of independent samples per duplicated participants for each and all cohorts.

#### What was your approach?

1. Setting seed + Randomizing the input histology file to remove selection bias.
2. Instead of using `sample`, I used `dplyr::distinct` to take the first occurrence of duplicated `Kids_First_Participant_ID`.

#### What GitHub issue does your pull request address?

https://github.com/PediatricOpenTargets/ticket-tracker/issues/135

#### Which areas should receive a particularly close look?

Code/README

#### Is there anything that you want to discuss further?

N/A

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes

#### What types of results are included (e.g., table, figure)?

table

#### What is your summary of the results?

N/A

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

File: analyses/independent-samples/util/independent_rna_samples.R [hunk 1]
```diff
@@ -134,48 +134,49 @@ independent_rna_samples <- function(independent_dna_sample_df,
...
+      
+      # deal with cancer group is NA to avoid missing samples
+      if(is.na(cancer_group_name)){
+        filtered_df <- sample_df %>% 
+          dplyr::filter(cohort == cohort_name) %>%
+          dplyr::filter(is.na(cancer_group))
+      }else{
+        filtered_df <- sample_df %>% 
+          dplyr::filter(cohort == cohort_name) %>%
+          dplyr::filter(cancer_group == cancer_group_name)
```
Comment (runjin326) [inline analyses/independent-samples/util/independent_rna_samples.R:160]: Thanks for cleaning this code up ;)

File: analyses/independent-samples/util/independent-samples.R [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  }
+  
+  
+  if(independent_level == "all-cohorts"){
+
+    independent_all <- sample_df %>%
+      dplyr::distinct(Kids_First_Participant_ID, .keep_all = TRUE) %>%
+      dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID)
+    
+    return(independent_all)
```
Comment (runjin326) [inline analyses/independent-samples/util/independent-samples.R:111]: Good idea to call each first and then generate `all-cohorts` based off that!

File: analyses/independent-samples/README.md [hunk 3]
```diff
@@ -1,5 +1,45 @@
...
+│   ├── independent-specimens.wgs.relapse.tsv
+│   ├── independent-specimens.wgswxspanel.primary-plus.eachcohort.tsv
+│   ├── independent-specimens.wgswxspanel.primary-plus.tsv
+│   ├── independent-specimens.wgswxspanel.primary.eachcohort.tsv
+│   ├── independent-specimens.wgswxspanel.primary.tsv
+│   ├── independent-specimens.wgswxspanel.relapse.eachcohort.tsv
+│   └── independent-specimens.wgswxspanel.relapse.tsv
+├── run-independent-samples.sh
+└── util
+    ├── independent-samples.R
```
Comment (runjin326) [inline analyses/independent-samples/README.md:39]: Thanks for adding this and reorganizing the codes!

**Review by runjin326 — APPROVED** [review]
Look good to me! Thanks for making the modifications!

File: analyses/independent-samples/util/independent_rna_samples.R [hunk 4]
```diff
@@ -134,48 +134,49 @@ independent_rna_samples <- function(independent_dna_sample_df,
...
       }
+      
+      # find the independent samples for the specific cancer group and cohort
+      independent_filtered <- filtered_df %>%
+        dplyr::distinct(Kids_First_Participant_ID, .keep_all = TRUE) %>%
+        dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID)
```
Comment (logstar) [inline analyses/independent-samples/util/independent_rna_samples.R:168]:
```suggestion
      # "If there are multiple rows for a given combination of inputs, only the first
      # row will be preserved. If omitted, will use all variables." -- distinct in dplyr 0.8.3
      independent_filtered <- filtered_df %>%
        dplyr::distinct(Kids_First_Participant_ID, .keep_all = TRUE) %>%
        dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID)
```

File: analyses/independent-samples/util/independent-samples.R [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        # get cancer groups that match to cancer_group_name
+        filtered_df <- sample_df %>% 
+          dplyr::filter(cohort == cohort_name) %>%
+          dplyr::filter(cancer_group == cancer_group_name)
+      }
+      
+      # find the independent samples for the specific cancer group and cohort
+      independent_filtered <- filtered_df %>%
+        dplyr::distinct(Kids_First_Participant_ID, .keep_all = TRUE) %>%
+        dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID)
```
Comment (logstar) [inline analyses/independent-samples/util/independent-samples.R:96]:
```suggestion
      # "If there are multiple rows for a given combination of inputs, only the first
      # row will be preserved. If omitted, will use all variables." -- distinct in dplyr 0.8.3
      independent_filtered <- filtered_df %>%
        dplyr::distinct(Kids_First_Participant_ID, .keep_all = TRUE) %>%
        dplyr::select(Kids_First_Participant_ID, Kids_First_Biospecimen_ID)
```

File: analyses/independent-samples/util/independent-samples.R [hunk 6]
```diff
@@ -0,0 +1,87 @@
+independent_samples <- function(histology_df,
```
Comment (logstar) [inline analyses/independent-samples/util/independent-samples.R]:
```suggestion
# independent_samples.R

#' Generate a vector of unique samples
#' 
#' The samples from this function will be unique with respect to participants
#' i.e. only no two samples will come from the same participant. The input list 
#' should be pre-filtered by `experimental_strategy` and `sample_type`.  
#' 
#' 
#' @param histology_df A data frame of samples, with columns corresponding to those
#'   in `histologies.tsv`
#' @param independent_level Designates whether we want to count independent samples in 
#'  different cohorts as independent or not. "all-cohorts" consider the same sampe
#'  in different cohorts as the same sample and "each-cohort" consider the same sample
#'  in different cohorts as "independent" (different). 
#' @param tumor_types Designates which types of tumors will be included. Options
#'   are "primary" to include only primary tumors, "prefer_primary" to include
#'   primary tumors when available, but fall back to other types, or "any" to
#'   randomly select among all available specimens. As of v6, primary tumors
#'   are defined as those designated "Initial CNS Tumor" or "Primary Tumor" in the
#'   `tumor_descriptor` field.
#' @param seed An optional random number seed. 
#' 
#' @return a data frame of Participant and Specimen IDs, each present only once.
independent_samples <- function(histology_df, 
```

Even though this is outdated, it was helpful when I first went over this module.

**Review by logstar — APPROVED** [review]
Thank you for preparing this PR @komalsrathi ! Again, awesome solution to issue <https://github.com/PediatricOpenTargets/ticket-tracker/issues/135>!

The `run-independent-samples.sh` runs well, and all `results/*tsv` are reproduced identically. The numbers in the added `03-qc-independent-samples.nb.html` are also very informative.

Following are some minor suggestions on additional comments.

**Comment by komalsrathi** [comment]
Thanks both! I have committed the suggestions by @logstar as well. Please feel free to merge unless there is anything amiss.

**Pull request merged**